# Grouping Data and Aggregating Variables

    1. Import libraries and data set
    2. Aggregating data with agg() function
        A. Derive a new column, loyalty flag
        B. Create a new column, loyalty flag, with transform() and loc()
    3. Looking at spending habits of different customers
        A. Checking basic statistics
        B. Looking at loyalty flag as percentage of whole
        C. Investigating outlier values
    4. Creating a spending flag
        A. Create spending flag column
        B. Assign users based on spending habits
    5. Creating an order frequency flag
        A. Create order freq flag column
        B. Assign users based on order frequency habits

# 1. Import libraries and data set

In [1]:
# import libraties
import pandas as pd
import numpy as np
import os

In [2]:
#folder path into usable string
path = r'/Users/manuellituma/01-2023 Instacart Basket Analysis'

In [4]:
#import ords_prods_merge dataset
df = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'merged_new_variables.pkl'))

In [5]:
#check dataframe output
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day busiest_period_of_day  
0  Regularly busy        Average orders  
1  Regularly busy        Average orders  
2  Regularly busy           Most orders  
3      Least busy        Average orders  
4      Least busy           Most orders

In [7]:
#checking out put 
print('Shape of new dataframe:', df.shape)

Shape of new dataframe: (32404859, 17)


In [8]:
print('Sample of new dataframe')
df.sample(5)

Sample of new dataframe


order_id  user_id  order_number  orders_day_of_week  \
17356565   1375488    98126            22                   5   
11290126   1679677   156817            23                   1   
5027633     158389    61242             4                   3   
19227106    649576    56886            87                   5   
1635937    1111717    46633             5                   5   

          order_hour_of_day  days_since_prior_order  product_id  \
17356565                 13                     1.0       41750   
11290126                 13                     4.0       14678   
5027633                  10                    14.0       39275   
19227106                 11                     1.0         329   
1635937                  12                    11.0       24852   

          add_to_cart_order  reordered _merge                  product_name  \
17356565                  1          0   both           Gluten Free Fusilli   
11290126                  7          1   both           Organic Frozen Peas   
5027633                   9          0   both           Organic Blueberries   
19227106                  1          1   both  Organic Whole Grassmilk Milk   
1635937                   2          1   both                        Banana   

          aisle_id  department_id  prices    price_range_loc     busiest_day  \
17356565        12              9     7.5  Mid-range product  Regularly busy   
11290126       116              1     1.3  Low_range product  Regularly busy   
5027633        123              4     8.3  Mid-range product  Regularly busy   
19227106        84             16     7.5  Mid-range product  Regularly busy   
1635937         24              4    12.3  Mid-range product  Regularly busy   

         busiest_period_of_day  
17356565           Most orders  
11290126           Most orders  
5027633            Most orders  
19227106           Most orders  
1635937            Most orders

# 2. Aggregating data with the agg() function

Aggregating the mean of the "order_number" column grouped by "department_id" for the entire dataframe. This will give us the minimum, maximum, and mean of orders by department.

In [9]:
# group df_ords_prods_merge by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df:')
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df:


order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

### Creating a subset to compare aggregations against the entire dataframe

In [10]:
#creating subset and printing to check output
df_subset = df[:1000000]

print('First five results of df_ords_prods_merge subset')
df_subset.head(5)

First five results of df_ords_prods_merge subset


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day busiest_period_of_day  
0  Regularly busy        Average orders  
1  Regularly busy        Average orders  
2  Regularly busy           Most orders  
3      Least busy        Average orders  
4      Least busy           Most orders

In [11]:
# group df_subset by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df_subset:')
df_subset.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df_subset:


order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

### Observations:

Whilst there are some differences between the aggregations seen in the subset compared to the entire dataframe, they tend to be within reasonable ranges. I calculated the percentage change for four departments: department 1 was within 4%, department 8 was within 7%, department was within 2%, and department 21 was within 4%.

There is some variance in the max values in the subset. The max values for all departments in the entire dataset is 99, however, in some of the subset department it is lower, for example, department 8 is only 91. Again, this is not unreasonable to expect when only looking at a slice of the data.

# 3. Creating a loyalty flag for customers using transform() and loc() functions

The client would like to better understand the loyalty status of their customer base, i.e., the extent to which customers return to purchase from Instacart. We are going to column which calculates the maximum number of orders per user, and then create a flag which groups users into one of three categories based on how many orders they have made.

# A. Creating a new column, 'max_order' to place the results of the maximum orders for each user

In [12]:
# create a new column 'max_order2' containing the maximum order number for each user_id, calculated using the groupby() and transform() functions
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

In [13]:
#check output by looks at user_id and max_order
print('Sample user_id and max_order output:')
df[['user_id', 'max_order']].sample(10)

Sample user_id and max_order output:


user_id  max_order
25128210   190200         35
1126965    169683         35
12617164   110521          6
9762170     56550         36
24391381   147271         59
31534488   195150          6
22040783    49973         77
2220116     56629         18
12868144   131514         11
20811908   125913         47

In [14]:
df.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
16744616    255724    93471            48                   0   
253556      275145   102765             7                   0   
10456943   2061081    52177            57                   6   
4056391    2265903    76854            41                   6   
26806539    600617    75287            15                   5   

          order_hour_of_day  days_since_prior_order  product_id  \
16744616                 14                     6.0       36011   
253556                   15                    21.0       13176   
10456943                 12                     8.0        6184   
4056391                  17                     6.0       43961   
26806539                 16                     0.0       49192   

          add_to_cart_order  reordered _merge  \
16744616                  7          1   both   
253556                    5          1   both   
10456943                  1          1   both   
4056391                  14          0   both   
26806539                  8          1   both   

                               product_name  aisle_id  department_id  prices  \
16744616              Organic Fat Free Milk        84             16     3.2   
253556               Bag of Organic Bananas        24              4    10.3   
10456943                        Clementines        32              4     4.3   
4056391   Organic Peeled Whole Baby Carrots       123              4     3.8   
26806539     Holler Mountain Organic Coffee        26              7     9.3   

            price_range_loc     busiest_day busiest_period_of_day  max_order  
16744616  Low_range product     Busiest day           Most orders         55  
253556    Mid-range product     Busiest day           Most orders          7  
10456943  Low_range product  Regularly busy           Most orders         87  
4056391   Low_range product  Regularly busy        Average orders         53  
26806539  Mid-range product  Regularly busy           Most orders         27

# B. Derive a new column, 'loyalty_flag' using the loc() function and the values found in "max_order"

Creating a flag which will assign each customer into one of three categores based on how many times they have purchased from Instacart. This loyalty flag allows us to group our customers by their loyalty status.

In [15]:
df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [16]:
df.loc[df['max_order'] <= 40, 'loyalty_flag'] = 'Regular customer'

In [17]:
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
#checking ouput of flag creation
df['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [19]:
# checking the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column
print('Sample output from columns user_id, max_order and loyalty_flag columns:')
df[['user_id', 'max_order', 'loyalty_flag']].sample(20)

Sample output from columns user_id, max_order and loyalty_flag columns:


user_id  max_order      loyalty_flag
23462150    35948         38  Regular customer
16784903   188217         11  Regular customer
29475477   139234          7      New customer
18161123   157041         35  Regular customer
30228809   173293         32  Regular customer
6073226     83299         15  Regular customer
3432042     16053          6      New customer
3343544    114883         39  Regular customer
11712086     1704         47    Loyal customer
7275572    156986         24  Regular customer
32065923    49613         33  Regular customer
10060976   117676         12  Regular customer
31768533   121984         67    Loyal customer
22999944    54012         44    Loyal customer
24847449    94497         34  Regular customer
7137556     77653         20  Regular customer
10030050   116472         37  Regular customer
24620187   174110         14  Regular customer
307783     133033          5      New customer
4921283    137000         27  Regular customer

# 4. Looking at the spending habits of different customers

Looking at how loyalty status impacts spending habits.


# A. Check basic statistics of product prices for each loyalty category (loyal, regular, and new), and assess whether spending habits differ across the different groups.

In [20]:
#basic statistical comparison of 'loyalty_flag' column, followed by calculation of mean, min, max, and sum of the 'prices' column
print('The mean, min, max, and sum of prices grouped by the loyalty_flag:')

df.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'sum']})

The mean, min, max, and sum of prices grouped by the loyalty_flag:


prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2

In [21]:
#confirming the numbers of each type of customer
df.groupby('loyalty_flag')['user_id'].count()

loyalty_flag
Loyal customer      10284093
New customer         6243990
Regular customer    15876776
Name: user_id, dtype: int64

# B. Calculating each loyalty flag as a percentage of total orders

In [22]:
#setting loyalty_flag as variable 
loyalty_flag = df.loyalty_flag

In [23]:
#calculating loyalty_flag as % of whole
percent_loyalty_flag = loyalty_flag.value_counts(normalize = True)
print(percent_loyalty_flag)

Regular customer    0.489950
Loyal customer      0.317363
New customer        0.192687
Name: loyalty_flag, dtype: float64


In [24]:
#returning results in easier to read format 
percent_loyalty_flag_100 = loyalty_flag.value_counts(normalize = True).mul(100).round(1).astype(str) + '%'
print(percent_loyalty_flag_100)

Regular customer    49.0%
Loyal customer      31.7%
New customer        19.3%
Name: loyalty_flag, dtype: object


### Observations

Almost 50% of orders fall within the regular customer category, just over 30% are from loyal customers, and just under 20% are new customers. Interestingly, the average price for loyal customers is the lowest of the three groups whilst new customers have the highest which I was surprised by. Some interesting futher analysis could be focused on the types of products loyal customers are buying.

The maximum price value of $99,999 seems like an extreme outlier, and requires further investigation. Significant outliers like this could skew our data.

# C. Investigating the outlier prices

In [25]:
#returning a list of all prices, ordered by most expensive
df['prices'].drop_duplicates().sort_values(ascending = False)

29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

##### There are two prices which appear to be outliers: 99,999 and 14,900

In [26]:
#filtering the dataset by two outlier prices
max_price = df.loc[df['prices'].isin([99999, 14900])]

In [27]:
max_price.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
10033276   2981940   138594            22                   6   
10031291   3030621    49175            18                   1   
29165886    898381   118950             1                   2   
10030974     89861    30348             4                   3   
10034315   1793763   186148             4                   3   

          order_hour_of_day  days_since_prior_order  product_id  \
10033276                 19                     6.0       21553   
10031291                  8                    10.0       21553   
29165886                  9                     NaN       33664   
10030974                 13                    30.0       21553   
10034315                 13                     2.0       21553   

          add_to_cart_order  reordered _merge  \
10033276                  9          0   both   
10031291                 10          0   both   
29165886                  2          0   both   
10030974                 12          1   both   
10034315                  2          0   both   

                              product_name  aisle_id  department_id   prices  \
10033276  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10031291  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
29165886             2 % Reduced Fat  Milk        84             16  99999.0   
10030974  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10034315  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   

             price_range_loc     busiest_day busiest_period_of_day  max_order  \
10033276  High-range product  Regularly busy        Average orders         26   
10031291  High-range product  Regularly busy        Average orders         20   
29165886  High-range product  Regularly busy           Most orders         32   
10030974  High-range product  Regularly busy           Most orders         14   
10034315  High-range product  Regularly busy           Most orders          5   

              loyalty_flag  
10033276  Regular customer  
10031291  Regular customer  
29165886  Regular customer  
10030974  Regular customer  
10034315      New customer

In [28]:
max_price.shape

(5127, 19)

### Observations

There are 5127 rows impacted by these outlier prices of 99,999 and 14,900. Ideally I would conduct some more investigation to see if we can figure out the correct price and amend the dataset to reflect. Results including these values as they currently are may impact the mean values returned in the basic statistics returned above.

# 5. Create a spending flag for each user based on the average price across all their order

The client would like to better understand how customers are spending money.

If the mean of the prices of products purchased by a user is lower than 10, then flag them as a "Low spender".

If the mean of the prices of products purchased by a user is higher than or equsl to 10, then flag them as a "High spender"

# A. Create new column called spending flag

In [29]:
#create a new columns called "spending_flag"
df['avg_price'] = df.groupby(['user_id'])['prices'].transform(np.mean)

In [30]:
#round the averages
df['avg_price'] = df.groupby(['user_id'])['prices'].transform(np.mean).round(2)

In [31]:
#check output
print('Sample of user_id and average price output:')
df[['user_id', 'avg_price']].sample(10)

Sample of user_id and average price output:


user_id  avg_price
23661977   180367       7.67
9043781    118919       9.58
11161472   146061       8.19
21545773    44676       7.29
13622146    74261       8.13
2674267     38766       7.86
23995619   159999       8.70
28367468    76044       7.50
8399875    162079       7.16
6351089    164723       7.93

# B. Create a flag which tags any user with an average spend of more than or equal to 10, as "high spender". Any customers with an average spend of less than 10 will be marked as "Low spender"

In [32]:
df.loc[df['avg_price'] >= 10, 'spending_flag']='High spender'

In [33]:
df.loc[df['avg_price'] < 10, 'spending_flag']='Low spender'

In [34]:
#check output
df.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
31195026   2179016   152960             5                   5   
24307798   2097668   131668             4                   3   
22151458    675339   104540             6                   1   
32186827   2984564    94951             6                   3   
25932841   2647326    49580             7                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
31195026                 12                    13.0       41137   
24307798                 17                     3.0       13819   
22151458                 10                    10.0        2717   
32186827                 15                    10.0       37302   
25932841                  9                     9.0        2380   

          add_to_cart_order  reordered _merge  ... aisle_id  department_id  \
31195026                  4          0   both  ...       34              1   
24307798                  3          1   both  ...      117             19   
22151458                  8          0   both  ...      115              7   
32186827                  3          1   both  ...       61             19   
25932841                  5          0   both  ...      104             13   

          prices    price_range_loc     busiest_day busiest_period_of_day  \
31195026     2.2  Low_range product  Regularly busy           Most orders   
24307798     5.2  Mid-range product  Regularly busy        Average orders   
22151458     9.2  Mid-range product  Regularly busy           Most orders   
32186827     6.2  Mid-range product  Regularly busy           Most orders   
25932841    11.3  Mid-range product  Regularly busy           Most orders   

         max_order      loyalty_flag avg_price  spending_flag  
31195026        13  Regular customer      7.80    Low spender  
24307798        59    Loyal customer      7.81    Low spender  
22151458        38  Regular customer      8.30    Low spender  
32186827        16  Regular customer      7.41    Low spender  
25932841        17  Regular customer      8.72    Low spender  

[5 rows x 21 columns]

# 6. Create an order frequency flag which corresponds to the regularity of a customer's order behaviour

The client would like to better understand the frequency of customer purchases

If the median of "days_since_prior_order" is higher than 20, the customer should be labeled a "Non-frequent customer"

If the median of "days_since_prior_order" is higher than 10 and lower than or equal to 20, then the customer should be labeled a "Regular customer"

If the median is lower than or equal to 10, then the customer should be labeled a "Frequent customer"

In [35]:
#create a new column "order_freq" containing the median in days_since_prior_order
df['order_freq'] = df.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [36]:
# check the output of 'user_id' and 'order_freq'
print('Sample user_id and order_freq output:')
df[['user_id', 'order_freq']].sample(10)

Sample user_id and order_freq output:


user_id  order_freq
25942182   198780        26.0
10917823    86422        11.0
10590510    36118        14.0
11108359    46113        26.0
17766708    27708        15.0
10186598   132182        25.0
24098198    87880         4.0
29659286   144104         7.0
23617584    65156        16.0
12299815   124372        11.0

#### Create a flag which tags any user with a median frequency order of less than or equal to 10 as "non-frequent customer", higher than 10 and lower than or equal to 20 as "regular customer", and higher than 20 as "frequent customer"

In [37]:
# create the 'order_freq_flag' column based on each user's median in days_since_prior_order
df.loc[df['order_freq'] > 20, 'order_freq_flag'] = 'Non-frequent customer'

In [38]:
df.loc[(df['order_freq'] > 10) & (df['order_freq'] <= 20), 'order_freq_flag'] = 'Regular customer'

In [39]:
df.loc[df['order_freq'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [40]:
#check the output and count of each new flag

print('Count of each order frequency variable')
df['order_freq_flag'].value_counts(dropna = False)

Count of each order frequency variable


Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_freq_flag, dtype: int64

# Export Dataset

In [41]:
df.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'full_merged5.pkl'))